### PR TITLE
[refactor](opt) optimize BE code of array_distinct function

### PR DESCRIPTION
**Pull request by superdiaodiao**
## Proposed changes

Issue Number: close #xxx

refactor part of BE code of `array_distinct` function by:
1. optimize the head files
2. use type deduction(auto)
3. use macro definition to simplify if-else code
4. add error log for unsupported element type of array

### CODE REVIEW

**Comment by superdiaodiao** [comment]
run buildall

**Comment by superdiaodiao** [comment]
run buildall

File: be/src/vec/functions/array/function_array_distinct.h [hunk 1]
```diff
@@ -257,63 +254,38 @@ class FunctionArrayDistinct : public IFunction {
                           IColumn& dest_column, ColumnArray::Offsets64& dest_offsets,
                           const NullMapType* src_null_map, NullMapType* dest_null_map,
                           DataTypePtr& nested_type) const {
+#define EXECUTE_NUMBER(TYPE, NAME)                                                      \
+    if (which.is_##NAME()) {                                                            \
+        res = _execute_number<TYPE>(src_column, src_offsets, dest_column, dest_offsets, \
```
Comment (cambyzju) [inline be/src/vec/functions/array/function_array_distinct.h:259]:
```suggestion
        return _execute_number<TYPE>(src_column, src_offsets, dest_column, dest_offsets, \
```

**Comment by superdiaodiao** [comment]
run buildall

File: be/src/vec/functions/array/function_array_distinct.h (hunk 1 above)
Comment (superdiaodiao) [reply be/src/vec/functions/array/function_array_distinct.h]: done

**Comment by superdiaodiao** [comment]
run p0
run p1

**Comment by superdiaodiao** [comment]
run p0, run p1

**Comment by superdiaodiao** [comment]
run p0

**Comment by superdiaodiao** [comment]
run p1

**Comment by superdiaodiao** [comment]
run p0

**Comment by superdiaodiao** [comment]
run p1

**Comment by superdiaodiao** [comment]
run cloud_p0, run cloud_p1

**Comment by superdiaodiao** [comment]
run cloud_p0

**Comment by superdiaodiao** [comment]
run cloud_p1

**Comment by superdiaodiao** [comment]
run buildall

**Comment by superdiaodiao** [comment]
run buildall

**Review by lide-reed — APPROVED** [review]
LGTM

File: be/src/vec/functions/array/function_array_distinct.h [hunk 2]
```diff
@@ -257,63 +254,38 @@ class FunctionArrayDistinct : public IFunction {
...
+        EXECUTE_NUMBER(ColumnDecimal256, decimal256)
+        EXECUTE_NUMBER(ColumnDecimal128V2, decimal128v2)
+        if (which.is_string()) {
             res = _execute_string(src_column, src_offsets, dest_column, dest_offsets, src_null_map,
                                   dest_null_map);
         }
+
+#undef EXECUTE_NUMBER
+
         return res;
```
[the remark this answers is not in the document]
Comment (superdiaodiao) [reply be/src/vec/functions/array/function_array_distinct.h]: Yes, make sense.

fixed

File: be/src/vec/functions/array/function_array_distinct.h [hunk 3]
```diff
@@ -257,63 +254,38 @@ class FunctionArrayDistinct : public IFunction {
                           IColumn& dest_column, ColumnArray::Offsets64& dest_offsets,
                           const NullMapType* src_null_map, NullMapType* dest_null_map,
                           DataTypePtr& nested_type) const {
+#define EXECUTE_NUMBER(TYPE, NAME)                                                      \
+    if (which.is_##NAME()) {                                                            \
+        res = _execute_number<TYPE>(src_column, src_offsets, dest_column, dest_offsets, \
+                                    src_null_map, dest_null_map);                       \
+    } else
+
```
[the remark this answers is not in the document]
Comment (superdiaodiao) [reply be/src/vec/functions/array/function_array_distinct.h:262]: OK.

File: be/src/vec/functions/array/function_array_distinct.h [hunk 4]
```diff
@@ -254,13 +254,12 @@ class FunctionArrayDistinct : public IFunction {
...
                           DataTypePtr& nested_type) const {
-#define EXECUTE_NUMBER(TYPE, NAME)                                                      \
-    if (which.is_##NAME()) {                                                            \
-        res = _execute_number<TYPE>(src_column, src_offsets, dest_column, dest_offsets, \
-                                    src_null_map, dest_null_map);                       \
+#define EXECUTE_NUMBER(TYPE, NAME)                                                       \
+    if (which.is_##NAME()) {                                                             \
+        return _execute_number<TYPE>(src_column, src_offsets, dest_column, dest_offsets, \
+                                     src_null_map, dest_null_map);                       \
     } else
```
[the remark this answers is not in the document]
Comment (superdiaodiao) [reply be/src/vec/functions/array/function_array_distinct.h]: fixed

File: be/src/vec/functions/array/function_array_distinct.h [hunk 5]
```diff
@@ -257,64 +254,41 @@ class FunctionArrayDistinct : public IFunction {
...
-        } else if (which.is_decimal256()) {
-            res = _execute_number<ColumnDecimal256>(src_column, src_offsets, dest_column,
-                                                    dest_offsets, src_null_map, dest_null_map);
-        } else if (which.is_decimal128v2()) {
-            res = _execute_number<ColumnDecimal128V2>(src_column, src_offsets, dest_column,
-                                                      dest_offsets, src_null_map, dest_null_map);
-        } else if (which.is_string()) {
-            res = _execute_string(src_column, src_offsets, dest_column, dest_offsets, src_null_map,
-                                  dest_null_map);
+        EXECUTE_NUMBER(ColumnUInt8, uint8)
```
[the remark this answers is not in the document]
Comment (superdiaodiao) [reply be/src/vec/functions/array/function_array_distinct.h]: added

**Comment by superdiaodiao** [comment]
run cloud_p0

**Comment by superdiaodiao** [comment]
run cloud_p0

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**